services/src/test/java/org/keycloak/utils/TimePeriodFormatterTest.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2018 Red Hat Inc. and/or its affiliates and other contributors
+ * as indicated by the @authors tag. All rights reserved.
+ */
+package org.keycloak.utils;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * Unit test for {@link TimePeriodFormatter}
+ *
+ * @author Vlastimil Elias (velias at redhat dot com)
+ */
+public class TimePeriodFormatterTest {
+
+    @Test
+    public void second_zero() {
+        TimePeriodFormatter tested = new TimePeriodFormatter(0);
+        assertFormat(tested, 0, "seconds");
+    }
+
+    @Test
+    public void second_one() {
+        TimePeriodFormatter tested = new TimePeriodFormatter(1);
+        assertFormat(tested, 1, "second");
+    }
+
+    @Test
+    public void second_more() {
+        TimePeriodFormatter tested = new TimePeriodFormatter(2);
+        assertFormat(tested, 2, "seconds");
+        tested = new TimePeriodFormatter(5);
+        assertFormat(tested, 5, "seconds");
+        tested = new TimePeriodFormatter(24);
+        assertFormat(tested, 24, "seconds");
+        tested = new TimePeriodFormatter(59);
+        assertFormat(tested, 59, "seconds");
+        tested = new TimePeriodFormatter(61);
+        assertFormat(tested, 61, "seconds");
+    }
+
+    @Test
+    public void minute_one() {
+        TimePeriodFormatter tested = new TimePeriodFormatter(60);
+        assertFormat(tested, 1, "minute");
+    }
+
+    @Test
+    public void minute_more() {
+        TimePeriodFormatter tested = new TimePeriodFormatter(2 * 60);
+        assertFormat(tested, 2, "minutes");
+        tested = new TimePeriodFormatter(5 * 60);
+        assertFormat(tested, 5, "minutes");
+        tested = new TimePeriodFormatter(24 * 60);
+        assertFormat(tested, 24, "minutes");
+        tested = new TimePeriodFormatter(59 * 60);
+        assertFormat(tested, 59, "minutes");
+        tested = new TimePeriodFormatter(61 * 60);
+        assertFormat(tested, 61, "minutes");
+    }
+
+    @Test
+    public void hour_one() {
+        TimePeriodFormatter tested = new TimePeriodFormatter(60 * 60);
+        assertFormat(tested, 1, "hour");
+    }
+
+    @Test
+    public void hour_more() {
+        TimePeriodFormatter tested = new TimePeriodFormatter(2 * 60 * 60);
+        assertFormat(tested, 2, "hours");
```

Comment:
Please add the following case:
```java
TimePeriodFormatter tested = new TimePeriodFormatter(2 * 60 * 60 + 30 * 60);
assertFormat(tested, 2, "hours", 30, "minutes"); // <--- this needs some adjustments.
```

Reply:
Current implementation is aligned with Admin GUI, it also uses only one unit. 
These more complicated formats should bring even bigger burden to the i18n using current Freemarker templating engine. I'm really not sure if they should be in basic implementation which is OK for most cases.

Reply:
@hmlnarik BTW how should your example look like in English string? How should it look like in German, Czech? How to i18n this format to be flexible enough to accommodate different languages but keep i18n for simple languages (with one singular and one plural word per unit) as simple as small change in {{message.properties}} only which is currently in my PR?

Reply:
That is my point. I'd rather think this through before accepting implementation of this feature.
One way would be to introduce FreeMarker method e.g. `secondsToHumanReadable` (similarly to `msg`). So for example, the email-verification.ftl would read
```
${msg("emailVerificationBodyHtml", link, secondsToHumanReadable(linkExpiration), realmName)?no_esc}
```
There will be entry in messages_en.properties that would contain SPI name for the given language family
```
secondsToHumanReadable=default
```
The value of `secondsToHumanReadable` could then be `slavic`, `chinese` etc. The implementation of `secondsToHumanReadable` would look up the SPI by name (using `default` as fallback) and the other entries in the `messages_en.properties` would depend on this SPI requirements. For `default` implementation handling only singlular and plural, the entries would be e.g.:
```
timeUnitSecond=second
timeUnitSeconds=seconds
timeUnitMinute=minute
timeUnitMinutes=minutes
timeUnitHour=hour
timeUnitHour=hours
timeUnitDay=day
timeUnitDay=days
timeUnitSeparator=, 
```
The provider method would be "String getFormattedString(long seconds)" where it would convert the seconds into the human readable string using the properties above, joining the partial times via timeUnitSeparator.
This would be possible to use for English, German, French, etc. Similarly, for Slavic languages, there would be a single SPI with translations of "hour(s)", "minute(s)" etc. defined in the message_LL.properties. This would be possible to use for Czech, Slovak, Polish, Russian etc.

It is more generic and would work for any language family, usually without necessity for the user to implement their own SPI.

Reply:
yep, this proposal looks good. But I'd like to phase its implementation a bit to be able to get something into KC 3.x series to be available sooner for real use. I expect new SPI is show stopper for KC 3.x, so I'd like to update this PR to be implemented proposed way in ftl, but to have only simple formatter in java, without SPI/plugability - it is still great improvement for KC 3.x series. Then another future PR's targeted to KC 4.x only may improve formatter and introduce SPI/pluggability. WDYT?

Reply:
Getting this into 3.x is rather unlikely since payload for 3.x is already wrapped and the priority of this issue is minor. Rather than having an implementation that is known to be subject to change, let's make it properly in the first go.

Reply:
Is an SPI and this complication really needed? With the current approach there is single and plural. Is it really more complicated than that?

Reply:
Yes, grammar rules are not that simple in some languages.